### ISSUE TITLE
[Analyzer] Detect adding middleware services but middleware isn't used

### ISSUE THREAD

**Issue by JamesNK** [dotnet/aspnetcore#44243]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe the problem.

Developers can add the services for middleware, e.g. `AddCors`, and not realize that the middleware needs to be registered in middleware with `UseCors`. There is no error or indication to the developer about what they're missing.

### Describe the solution you'd like

An analyzer that looks at what middleware services are registered, what middleware is used in request pipeline, and warns the developer about middleware that's unused.

**Comment by davidfowl** [comment]
What happens if it's called by another method that the analyzer can't see the body of?

**Comment by JamesNK** [comment]
Suppress the warning?

Are you thinking of WebApplication which automatically adds UseAuthN and UseAuthZ to the middleware pipeline? The analyzer could detect WebApplication and infer that middleware is used.

**Comment by davidfowl** [comment]
Yes this is the only analyzer suggestion I think shouldn't be a warning by default. People will refactor things into methods and the analyzer will get confused quickly.

**Comment by JamesNK** [comment]
We might not necessarily be limited to detecting UseMiddleware calls in Main. We could check to see if there is an appropriate Use call anywhere in app code.

Depending how robust this is (i.e. minimal false positives) can determine the diagnostic level.

**Comment by davidfowl** [comment]
I'll feel better when I see how we plan to tackle that 😃
